### PR TITLE
11 implement parsergb and parsehdr for rgba128parser

### PR DESCRIPTION
**Pull request by ClemensU42**
Finished implementing `ParseRGB` and `ParseHDR`.

### CODE REVIEW

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 1]
```diff
@@ -141,9 +145,84 @@ bool ParseHex(out RGBA128 result)
...
+					IsChannelInRange(g) &&
+					IsChannelInRange(b) &&
+					IsChannelInRange(a))
+				{
+					result = new RGBA128(r / 255f, g / 255f, b / 255f, a / 255f);
+					return true;
+				}
+
+			}
+			return YieldError(out result);
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:171]: There is a method down below called `ConvertRGBA` that performs the same function.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 2]
```diff
@@ -141,9 +145,84 @@ bool ParseHex(out RGBA128 result)
...
+		{
+			int index = 0;
+
+			int r = 0, g = 0, b = 0, a = 255;
+
+			if (findNextInt(content, ref index, ref r) &&
+				findNextInt(content, ref index, ref g) &&
+				findNextInt(content, ref index, ref b))
+			{
+				if (findNextInt(content, ref index, ref a)) return YieldError(out result);
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:158]: Does this logic work out? The alpha channel is optional, so we should not yield and error if it is missing.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 3]
```diff
@@ -141,9 +145,84 @@ bool ParseHex(out RGBA128 result)
...
+				while (char.IsDigit(src[index + len]) && (index + len) < (src.Length - 1)) len++; // get the length of the number
+
+				if (len == 0) return false;
+
+				ReadOnlySpan<char> num = src.Slice(index, len);
+				result = int.Parse(num);
+				index += len;
+				return true;
+			}
+		}
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:188]: Small code style suggestions for this method:
- Method names should be `PascalCase` not `camelCase` (note the first uppercase)
- Do not use short abbreviations unless they make the lines _extremely_ long (ie. src -> source, len -> length)
- Do not put two one line statements into one line (ie. don't do `while (something) if (somethingElse) thenDoStuff`) because it gets confusing, add braces and put them on new lines. 
- It is probably better to rename `index` to `head` since we are scanning through a `ReadOnlySpan` and `index` is the 'head' of the scanning.
- It is probably better to record the end of the number rather than the length of it, since then you do not need to to `idnex + len` every time. And then after processing the current integer you can just do `index = end` (also renaming `index` to `head` and `end` to `tail` probably makes more sense, so record `int tail = head;` after your first while loop when you found the start of the integer, then increment the `ref head` to find the 'head' of the integer).
- Use `int.TryParse` rather than `int.Parse` because we do not want exceptions but rather a `bool` to report whether the conversion succeeded or not.
- `ReadOnlySpan` has range slicers, so you can do `int.TryParse(source[tail..head], out result)`.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 4]
```diff
@@ -141,9 +145,84 @@ bool ParseHex(out RGBA128 result)
...
+				ReadOnlySpan<char> num = src.Slice(index, len);
+				float tempRes = 0f;
+				bool parsed = float.TryParse(num, result: out tempRes);
+
+				if (parsed) result = tempRes;
+				else return false;
+				index += len;
+
+				return true;
+			}
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:224]: Please also revise this method based on the comment for the `findNextInt` method in `ParseRGB`.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 5]
```diff
@@ -76,63 +80,63 @@ bool ParseHex(out RGBA128 result)
...
+						if (ConvertHex(content[6], out int a0) && ConvertHex(content[7], out int a1))
+						{
+							a = CombineHex(a0, a1);
+							goto case 6;
+						}
 
-					goto default;
-				}
+						goto default;
+					}
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:139]: If you can, can you revert these changes with GitHub Desktop? :)

**Review by GaryHuan9 — CHANGES_REQUESTED** [review]
Nice work! Just some code style suggestions 👍

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 6]
```diff
@@ -171,56 +165,72 @@ bool ParseRGB(out RGBA128 result)
 			return YieldError(out result);
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:158]: You missed the comment about this code block being a duplicate of `ConvertRGBA` down below.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 7]
```diff
@@ -171,56 +165,72 @@ bool ParseRGB(out RGBA128 result)
...
 			{
-				int len = 0;
-
-				while (!char.IsDigit(src[index])) if (src.Length <= index++) return false; // find first digit
-				while (char.IsDigit(src[index + len]) && (index + len) < (src.Length - 1)) len++; // get the length of the number
+				if (alpha)
+				{
+					result = 255;
+					return true;
+				}
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:167]: Should not return this early; this would ignore the alpha channel if we actually have one.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 8]
```diff
@@ -171,56 +165,72 @@ bool ParseRGB(out RGBA128 result)
...
+				while (!char.IsDigit(source[head]))
+					if (source.Length <= head++)
+					{
+						result = 0;
+						return false; // find first digit
+					}
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:173]: Add braces to this `while` loop please.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 9]
```diff
@@ -171,56 +165,72 @@ bool ParseRGB(out RGBA128 result)
...
+				int tail = head;
+				while (char.IsDigit(source[tail]) && tail < (source.Length - 1)) tail++; // get the length of the number
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:175]: If you swap the compare order, we can simply this into:
```cs
while (tail < source.Length && char.IsDigit(source[tail])) tail++;
```

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 10]
```diff
@@ -171,56 +165,72 @@ bool ParseRGB(out RGBA128 result)
...
-				else return false;
-				index += len;
+				if (tail == head)
+				{
+					result = 0;
+					return false;
+				}
 
-				return true;
+				head = tail;
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:225]: Similar comments from above for this method.

**Review by GaryHuan9 — CHANGES_REQUESTED** [review]
We are getting closer! I think we might need to write some unit tests next.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs (hunk 6 above)
Comment (ClemensU42) [reply Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:158]: I fixed that in the newest commit

Comment (GaryHuan9) [reply Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:158]: Oh lol did I not see that?

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 11]
```diff
@@ -157,22 +157,22 @@ bool ParseRGB(out RGBA128 result)
...
+				{
 					if (source.Length <= head++)
 					{
 						result = 0;
 						return false; // find first digit
 					}
+				}
 				int tail = head;
-				while (char.IsDigit(source[tail]) && tail < (source.Length - 1)) tail++; // get the length of the number
+				while (tail < (source.Length - 1) && char.IsDigit(source[tail])) tail++; // get the length of the number
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:175]: No need the subtract by one I believe. Just doing `tail < source.Length` should work.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 12]
```diff
@@ -208,11 +208,13 @@ static bool findNextFloat(ReadOnlySpan<char> source, ref int head, out float res
...
 				while (!char.IsDigit(source[head]))
+				{
 					if (source.Length <= head++)
 					{
 						result = 0;
 						return false; // find first digit
 					}
+				}
 				int tail = head;
 				while ((char.IsDigit(source[tail]) || source[tail].Equals('.')) && tail < (source.Length - 1)) tail++; // get the length of the number
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:219]: Similar changes here needed, flip the logic and remove the `-1`

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 13]
```diff
@@ -157,22 +157,22 @@ bool ParseRGB(out RGBA128 result)
...
 					result = 255;
-					return true;
 				}
 				while (!char.IsDigit(source[head]))
+				{
 					if (source.Length <= head++)
 					{
 						result = 0;
 						return false; // find first digit
 					}
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:172]: This does not work correctly when `alpha` is `true`.

File: Echo.Core/src/Textures/Colors/RGBA128.Parser.cs [hunk 14]
```diff
@@ -157,22 +157,22 @@ bool ParseRGB(out RGBA128 result)
...
 			return YieldError(out result);
 
-			static bool IsChannelInRange(int channel) => channel > 0 && channel <= 255;
 			static bool FindNextInt(ReadOnlySpan<char> source, ref int head, out int result, bool alpha = false)
 			{
 				if (alpha)
 				{
 					result = 255;
-					return true;
 				}
```
Comment (GaryHuan9) [inline Echo.Core/src/Textures/Colors/RGBA128.Parser.cs:165]: Short `if` statements can be on one line.

**Review by GaryHuan9 — CHANGES_REQUESTED** [review]
Good progress, very close now. I think we should add some unit tests.
